feat(otlp-exporter-base): add retries

svetlanabrennan opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Adding retries to otlp http & proto exporters. 

Note: this pr does not add retries to the grpc otlp exporter.

Fixes #1233 

## Short description of the changes

Added retry logic to `sendWithHttp` and `sendWithXhr` functions. It does the following:

If the exporter gets a failed response that is retryable, we check if the response included a "retry-after" header. 
- If yes, retry again in the time specified in the retry-header which can be seconds or a date/time. 
- If no, we use an exponential backoff with jitter 

This logic is retried until there is a successful response or until the exporter timer times out (set via `timeoutMillis` in the exporter config or via env vars). If the exporter timers times outs, the request is aborted with a "Request Timeout" error. 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit Tests - for browser only

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

svetlanabrennan commented:
For some reason, using `req.destroy()` in a real http request export [here](https://github.com/open-telemetry/opentelemetry-js/blob/43320ee2e9ab1b85887ebe2ca2819f1a742c9c71/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L59) is not triggering any of the following events:

- 'aborted' on the res object
- 'error' on the res object with an error with message 'Error: aborted' and code 'ECONNRESET'.
- 'error' with an error with message 'Error: socket hang up' and code 'ECONNRESET'

These events should fire depending on when the request is aborted (before socket assigned, before connection succeeds or after response is received). 

If I use `req.abort()` then it works (fires the correct events) but `req.abort()` was deprecated in node 14 so I was trying to avoid using it in node 14 and higher. 

Any idea what I'm missing here or what's going on?

dyladan commented:
> For some reason, using `req.destroy()` in a real http request export [here](https://github.com/open-telemetry/opentelemetry-js/blob/43320ee2e9ab1b85887ebe2ca2819f1a742c9c71/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L59) is not triggering any of the following events:
> 
> * 'aborted' on the res object
> * 'error' on the res object with an error with message 'Error: aborted' and code 'ECONNRESET'.
> * 'error' with an error with message 'Error: socket hang up' and code 'ECONNRESET'
> 
> These events should fire depending on when the request is aborted (before socket assigned, before connection succeeds or after response is received).
> 
> If I use `req.abort()` then it works (fires the correct events) but `req.abort()` was deprecated in node 14 so I was trying to avoid using it in node 14 and higher.
> 
> Any idea what I'm missing here or what's going on?
> 
> If you run this [test](https://github.com/open-telemetry/opentelemetry-js/blob/43320ee2e9ab1b85887ebe2ca2819f1a742c9c71/experimental/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts#L592-L625) and use `req.destroy()` [here](https://github.com/open-telemetry/opentelemetry-js/blob/43320ee2e9ab1b85887ebe2ca2819f1a742c9c71/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L59) instead of `req.abort() and use node 16, you'll see what I'm referring to.

Those events only fire if you call `req.destroy` _after_ the response is received https://nodejs.org/api/http.html#httprequesturl-options-callback.

svetlanabrennan commented:
I would love some feedback on the current implementation of the retry logic. 

I still need to confirm and add the following items:
- "jitter" to the retry logic
- confirm the retryable codes (I think there is an issue open about this in the spec)

legendecas commented on experimental/packages/otlp-exporter-base/src/platform/browser/util.ts:
```diff
@@ -60,48 +66,86 @@ export function sendWithXhr(
-        const error = new OTLPExporterError(
-          'Request Timeout', xhr.status
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState === XMLHttpRequest.DONE && reqIsDestroyed === undefined) {
```
Can we initialize `reqIsDestroyed` with `false` instead?

svetlanabrennan replied:
Done

legendecas commented on experimental/packages/otlp-exporter-base/src/platform/node/util.ts:
```diff
@@ -69,63 +82,87 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
+      });
+
+      res.on('end', () => {
+        if (reqIsDestroyed === undefined) {
```
Ditto.

svetlanabrennan replied:
Done

legendecas reviewed (commented):
Overall LGTM, thank you for working on this!

svetlanabrennan commented:
I've added the following:

- jitter to exponential backoff
- added logic to check for "retry-header" to use in retry logic
- fix some tests and removed a flaky real http request test (node)

svetlanabrennan commented:
I'm unable to add real node tests to test this retry logic because the tests are very flaky when I use a real node http request. I tried to use "nock" but I found out that it overrides the client http request so I can't test the retry logic.

weyert commented:
> I'm unable to add real node tests to test this retry logic because the tests are very flaky when I use a real node http request. I tried to use "nock" but I found out that it overrides the client http request so I can't test the retry logic.

I am wondering if you could use `msw` for this?

legendecas commented on experimental/packages/otlp-exporter-base/src/platform/browser/util.ts:
```diff
@@ -59,49 +66,117 @@ export function sendWithXhr(
+  sendWithRetry();
+}
+
+function retrieveThrottleTime(retryAfter: string): number {
```
This parameter type looks abstruse to me. The comment below suggests that the `retryAfter` can either be a Date object or an integer, but the explicit type here suggests it is a string. 

The call site of this function pass in a value from `xhr.getResponseHeader()`, which should be a string. This seems like diverging from the mocks in the test.

svetlanabrennan replied:
@legendecas 

Well the `Retry-After` can be a a Date or an integer (representing seconds) and when I retrieve the Retry-After with `getResponseHeader()`, it returns a string so `retrieveThrottleTime` parameter is set to string type. But I'm able to use `typeof` inside `retrieveThrottleTime ` to check if `retryAfter` argument is an object or not. 

Maybe I can change `function retrieveThrottleTime(retryAfter: string)` to be `function retrieveThrottleTime(retryAfter: string | object)`. Do you have any recommendations on how else I can implement this logic? Appreciate the help.

legendecas replied:
What do you think about https://github.com/open-telemetry/opentelemetry-js/pull/3577/files#diff-67f1efb4b6dca82d1da8bd9d718dfca1b9126597c8e5b225a50f3926c9874d35R72?

svetlanabrennan replied:
This works when the retry-after is in seconds but it doesn't work when retry-after is a date. When the retry-after is a date, the `retrieveThrottleTime` won't work because `seconds` will be `NaN`.

Example:

```js
function retrieveThrottleTime(retryAfter?: string | null): number {
    if (retryAfter == null) {
      return -1;
    }

    console.log("retry after is ", retryAfter)
    // logs => Wed Feb 08 2023 11:53:14 GMT-0600 (Central Standard Time)

    const seconds = Number.parseInt(retryAfter, 10);
    console.log('seconds are ', seconds)
   // logs => NaN

    if (Number.isInteger(seconds)) {
      return seconds > 0 ? seconds * 1000 : -1;
    }
    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After#directives
    const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay * 1000;
    }
    return 0;
}
```

svetlanabrennan replied:
What about this small tweak to address the above issue:

```js
function retrieveThrottleTime(retryAfter?: string | null): number {
    if (retryAfter == null) {
      return -1;
    }

    const seconds = Number.parseInt(retryAfter, 10);

    if (isNaN(seconds)) {
      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After#directives
      const delay = (new Date(retryAfter).getTime() - Date.now()) / 1000;
      if (delay >= 0) {
        return delay * 1000;
      }
      return 0;
    } else {
      return seconds > 0 ? seconds * 1000 : -1;
    }
}
```

legendecas replied:
Do you referring to this test case? https://github.com/open-telemetry/opentelemetry-js/pull/3577/files#diff-8baa107c057bf551118789c29ad093d016a7ddb1a26a9c5ad02d67ff386e1dffR63

IIUC `Number.isInteger(NaN)` returns false so when `retry-after` response header is a date string, it would be parsed with `new Date` as suggested by the MDN doc.

svetlanabrennan replied:
Yes I'm referring to that test case. Thanks for the clarification on this:

> IIUC Number.isInteger(NaN) returns false so when retry-after response header is a date string, it would be parsed with new Date as suggested by the MDN doc.

I have a question on this part of the function:

```js
const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay * 1000;
    }
```

Shouldn't that be:

```js
const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay;
    }
```

Otherwise if we do `return delay * 1000` then the `delay` would be too large. For example, if the retry-after date is `new Date('2023-01-20T00:00:01.000Z')` and the now date is `new Date('2023-01-20T00:00:00.000Z')`, then this function will return the delay of `1000000` milliseconds which is 100 seconds but we need 1000 milliseconds for a 1 second delay for this test case.

legendecas replied:
Ah, yes, you're right. Thanks for pointing this out.

svetlanabrennan replied:
I've updated this function to match your feedback.

legendecas reviewed (commented):
Sorry for the delay.

I'm wondering if it would be possible to unify the retry handling around the `sendWithXhr` and `sendWithHttp` -- it only depends on the `'Retry-After'` header and the response status code. 

It may be simpler if the platform-dependent part can be minimized.

svetlanabrennan commented:
> I'm wondering if it would be possible to unify the retry handling around the sendWithXhr and sendWithHttp -- it only depends on the 'Retry-After' header and the response status code.

Sure I can give that a shot after thanksgiving break.

svetlanabrennan commented:
@legendecas Regarding your comment:
> I'm wondering if it would be possible to unify the retry handling around the `sendWithXhr` and `sendWithHttp` -- it only depends on the `'Retry-After'` header and the response status code.
> 
> It may be simpler if the platform-dependent part can be minimized.

Sorry for the delay on this.

I don't think I can unify the retry handling between `sendWithXhr` and `sendWithHttp`. There's quite a few difference between the two (ex. how the retry-after header is retrieve, how the `OTLPExporterError` is created, how the request status codes are retrieved....etc). Did you happen to see a way this can be done and point me in the right direction?

legendecas commented:
> ex. how the retry-after header is retrieve, how the OTLPExporterError is created, how the request status codes are retrieved....etc.

I think we should create a dedicated http client abstraction on both Node.js and browser, either by using widely adopted npm package (like [superagent](https://www.npmjs.com/package/superagent) or [axios](https://www.npmjs.com/package/axios))  or creating our own one (since we may need to tweak it to be able to use `sendBeacon` too). Anyway, I believe this can be done in a follow up PR.

svetlanabrennan commented:
I've updated the `parseRetryAfterToMillis` function based on @legendecas feedback. This PR is ready for review.

legendecas commented:
I'm going to merge this as @opentelemetry/instrumentation-http:test failure on node-windows-tests is being tracked at https://github.com/open-telemetry/opentelemetry-js/issues/3628.
